velox/functions/prestosql/registration/GeneralFunctionsRegistration.cpp

```diff
@@ -81,6 +81,7 @@ void registerAllSpecialFormGeneralFunctions() {
       Variadic<Generic<T1>>>({"in"});
   VELOX_REGISTER_VECTOR_FUNCTION(udf_concat_row, expression::kRowConstructor);
   registerIsNullFunction("is_null");
+  registerFailFunction({expression::kFail});
```

Comment:
Why do we register "fail" twice w/ and w/o prefix?

Reply:
Thanks. Please, document this change in the PR description.

Reply:
Thanks for the suggestion @mbasmanova, decoupled Presto `fail` function from `ExprOptimizer` as suggested by introducing lambda `failExpr`. The cast to right type is still added by the `ExprOptimizer` though since it has the context of failing expression's type. The `failExpr` lambda lets users define how the error message should be processed, such as by invoking Presto `fail` function. Is this fine?

Also added tests for `failExpr` and updated code documentation and PR description.

Reply:
Removed cast from `ExprOptimizer` and the onus is on the user to provide a `failExpr` with cast to valid type. Could you please confirm if this is fine or whether I should go back to previous approach where `ExprOptimizer` wraps `failExpr` with cast to expected type? cc: @aditi-pandit @czentgr

Reply:
@aditi-pandit, since the error message cannot be captured with a `LambdaTypedExpr`, a `TypedExpr` that replaces failing sub-expressions should suffice here. Could you please take a look at the fail expression handling? 

If more changes are needed for fail expression handling, would it be fine to work on this in a follow-up PR?